lib/kubeclient/informer.rb

```diff
@@ -78,37 +78,45 @@ def fill_cache
     end
 
     def watch_to_update_cache
-      @watcher = @client.watch_entities(@resource_name, watch: true, resource_version: @started)
-      stop_reason = 'disconnect'
-
-      # stop watcher without using timeout
-      @waiter = Thread.new do
-        sleep(@reconcile_timeout)
-        stop_reason = 'reconcile'
-        @watcher.finish
-      end
+      watcher_with_timeout do |watcher|
+        stop_reason = 'disconnect'
 
-      @watcher.each do |notice|
-        case notice[:type]
-        when 'ADDED', 'MODIFIED' then @cache[cache_key(notice[:object])] = notice[:object]
-        when 'DELETED' then @cache.delete(cache_key(notice[:object]))
-        when 'ERROR'
-          stop_reason = 'error'
-          break
-        else
-          @logger&.error("Unsupported event type #{notice[:type]}")
+        watcher.each do |notice|
+          case notice[:type]
+          when 'ADDED', 'MODIFIED' then @cache[cache_key(notice[:object])] = notice[:object]
+          when 'DELETED' then @cache.delete(cache_key(notice[:object]))
+          when 'ERROR'
+            stop_reason = 'error'
+            break
+          else
+            @logger&.error("Unsupported event type #{notice[:type]}")
+          end
+          @watching.each { |q| q << notice }
         end
-        @watching.each { |q| q << notice }
+
+        stop_reason
       end
-      @logger&.info("watch restarted: #{stop_reason}")
+    end
 
-      # wake the waiter unless it's dead so it does not hang around
-      begin
-        Thread.pass # make sure we get into the sleep state of the waiter
-        @waiter.run
-      rescue ThreadError # rubocop:disable Lint/SuppressedException
+    def watcher_with_timeout
+      @watcher = @client.watch_entities(@resource_name, watch: true, resource_version: @started)
+      timeout_deadline = Time.now + @reconcile_timeout
+      watcher_finished = false
+
+      finisher_thread = Thread.new do
+        sleep 0.5 until @stopped || watcher_finished || Time.now > timeout_deadline
```

Comment:
instead of using sleep as the condition for continuing (and using run as signalling), evaluate the conditions explicitly. this allows to wait for any of the different conditions - either stopped is triggered, the method ended, or the timeout expired.